### PR TITLE
[Kimi Perf] Fuse the three Kimi-K3 latent-MoE front projections, 1.1~1.4x kernel performance improvement

### PR DESCRIPTION
**Pull request by yewentao256**
## Purpose

Before

```bash
hidden_states [M, 7168]
 ├─ shared gate/up proj → [M, 1536]
 ├─ router proj         → [M, 896] → TopK
 └─ routed down proj    → [M, 3584]
```

Now

```bash
output = hidden_states @ merged_weight.T
shared, router, routed = output.split((1536, 896, 3584), dim=-1)
```

## Test

Acc covered in current unit test

Perf can be seen in this AI generated script

```bash
import torch
import triton

from vllm.model_executor.kernels.linear.cute_dsl.ll_bf16 import ll_bf16_gemm
from vllm.model_executor.layers.fused_moe.router.grouped_topk_router import (
    fused_grouped_topk,
)
from vllm.models.kimi_k3.nvidia.low_latency_gemm import try_low_latency_gemm
from vllm.models.kimi_k3.nvidia.model import _fused_moe_front

HIDDEN_SIZE = 7168
NUM_EXPERTS = 896
OUTPUT_SIZES = (1536, NUM_EXPERTS, 3584)
TOP_K = 16
TOKEN_COUNTS = (1, 2, 4, 8, 16, 32, 64, 128, 256)


def projection(hidden_states: torch.Tensor, weight: torch.Tensor) -> torch.Tensor:
    output = try_low_latency_gemm(hidden_states, weight)
    if output is not None:
        return output
    return torch.mm(hidden_states, weight.t())


def benchmark_token_count(
    num_tokens: int,
    merged_weight: torch.Tensor,
    correction_bias: torch.Tensor,
) -> None:
    shared_weight, router_weight, latent_weight = merged_weight.split(OUTPUT_SIZES)
    hidden_states = torch.randn(
        num_tokens,
        HIDDEN_SIZE,
        device=merged_weight.device,
        dtype=torch.bfloat16,
    )

    def topk_ids(router_logits: torch.Tensor) -> torch.Tensor:
        return fused_grouped_topk(
            hidden_states,
            router_logits,
            TOP_K,
            True,
            correction_bias,
            1,
            1,
            "sigmoid",
            1.0,
        )[1]

    def separate() -> torch.Tensor:
        projection(hidden_states, shared_weight)
        router_logits = (
            ll_bf16_gemm(hidden_states, router_weight)
            if num_tokens <= 16
            else torch.mm(hidden_states, router_weight.t(), out_dtype=torch.float32)
        )
        projection(hidden_states, latent_weight)
        return topk_ids(router_logits)

    def fused() -> torch.Tensor:
        _, router_logits, _ = _fused_moe_front(
            hidden_states, merged_weight, OUTPUT_SIZES
        )
        return topk_ids(router_logits)

    expected = separate()
    actual = fused()
    torch.cuda.synchronize()

    mismatch = expected.ne(actual).any(dim=1).float().mean().item()
    separate_ms = triton.testing.do_bench(separate, warmup=50, rep=200)
    fused_ms = triton.testing.do_bench(fused, warmup=50, rep=200)

    print(
        f"{num_tokens:6d}  {separate_ms * 1e3:11.3f}  "
        f"{fused_ms * 1e3:8.3f}  {separate_ms / fused_ms:7.3f}  "
        f"{mismatch:13.6f}"
    )


def main() -> None:
    torch.manual_seed(1)
    merged_weight = torch.randn(
        sum(OUTPUT_SIZES),
        HIDDEN_SIZE,
        device="cuda",
        dtype=torch.bfloat16,
    )
    merged_weight.mul_(HIDDEN_SIZE**-0.5)
    correction_bias = torch.randn(NUM_EXPERTS, device="cuda", dtype=torch.float32).mul_(
        0.01
    )

    print(f"GPU: {torch.cuda.get_device_name()}")
    print("tokens  separate_us  fused_us  speedup  topk_mismatch")

    for num_tokens in TOKEN_COUNTS:
        benchmark_token_count(num_tokens, merged_weight, correction_bias)


if __name__ == "__main__":
    main()

```

And we can get

```bash
tokens  separate_us  fused_us  speedup  topk_mismatch
     1       51.354    37.080    1.385       0.000000
     2       52.005    43.024    1.209       0.000000
     4       60.213    42.414    1.420       0.000000
     8       60.362    41.831    1.443       0.000000
    16       58.526    42.194    1.387       0.000000
    32       52.218    43.547    1.199       0.000000
    64       49.983    45.264    1.104       0.000000
   128       51.248    42.972    1.193       0.000000
   256       59.335    50.742    1.169       0.000000
```

### CODE REVIEW

File: vllm/models/kimi_k3/nvidia/model.py [hunk 1]
```diff
@@ -599,6 +612,12 @@ def __init__(
         self.use_mega_moe = (
             vllm_config.kernel_config.moe_backend == "deep_gemm_mega_moe"
         )
+        self._fused_front_weight: torch.Tensor | None = None
+        self._fused_front_config_eligible = (
```
Comment (mgoin) [inline vllm/models/kimi_k3/nvidia/model.py:616]: Do we use this "front" naming anywhere else? I haven't seen it before and it isn't obvious to me

File: vllm/models/kimi_k3/nvidia/model.py [hunk 2]
```diff
@@ -599,6 +612,12 @@ def __init__(
...
+            not use_sequence_parallel
+            and not vllm_config.parallel_config.enable_expert_parallel
+            and not vllm_config.model_config.enable_sleep_mode
+        )
```
Comment (mgoin) [inline vllm/models/kimi_k3/nvidia/model.py:620]: It seems this would also break weight reloading for RL and not work with quantization. This might have to be an opt-in feature at best unless we find a general solution... The "right way" would be to replace the modules in the model def itself with a specific merged module, not merge them after the fact manually

At the very least you should test this with a popular nvfp4 checkpoint used in our recipes like https://huggingface.co/RedHatAI/Kimi-K3-NVFP4 or https://huggingface.co/nvidia/Kimi-K3-NVFP4 that may quantize these layers

File: vllm/models/kimi_k3/nvidia/model.py [hunk 3]
```diff
@@ -135,6 +135,17 @@
 _ROUTED_DOWN_PROJ_STREAM_TOKEN_THRESHOLD = 256
 
 
+def _fused_moe_front(
+    hidden_states: torch.Tensor,
+    merged_weight: torch.Tensor,
+    sizes: tuple[int, int, int],
+) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+    output = torch.mm(hidden_states, merged_weight.t(), out_dtype=torch.float32)
```
Comment (mgoin) [inline vllm/models/kimi_k3/nvidia/model.py:143]: We have specialized kernels for bf16 gemm with fp32 output, although they might be specialized for routing. Would be worth trying to reuse the interface at least

File: vllm/models/kimi_k3/nvidia/model.py [hunk 4]
```diff
@@ -135,6 +135,17 @@
...
+    shared, router, routed = output.split(sizes, dim=-1)
+    dtype = hidden_states.dtype
+    return shared.to(dtype), router.contiguous(), routed.to(dtype)
```
Comment (mgoin) [inline vllm/models/kimi_k3/nvidia/model.py:146]: This is a bit sneaky that only the router output needs to be fp32 and the others are downcast, and that downcast is assumed for how we get contiguity as well. Worth a comment.

File: vllm/models/kimi_k3/nvidia/model.py [hunk 5]
```diff
@@ -835,6 +887,27 @@ def _router(
     def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
         num_tokens, hidden_size = hidden_states.shape
         hidden_states = hidden_states.view(-1, hidden_size)
+        down_proj = self.routed_expert_down_proj
+        if (
+            self._fused_front_weight is not None
+            and down_proj is not None
+            and 0 < num_tokens <= _ROUTED_DOWN_PROJ_STREAM_TOKEN_THRESHOLD
```
Comment (mgoin) [inline vllm/models/kimi_k3/nvidia/model.py:894]: How are we dynamically dispatching based on batch size if we already merged the weights?

**Review by mgoin — CHANGES_REQUESTED** [review]
Definitely needs more than unit testing since this is more than a kernel substitution. Please see my initial comments about concerns with quantization and weight reloading

File: vllm/models/kimi_k3/nvidia/model.py (hunk 4 above)
Comment (jeejeelee) [reply vllm/models/kimi_k3/nvidia/model.py:146]: IIUC, we need not downcast here
- shared can be consumed by `SituAndMul`
- routed can be consumed by quant ops

File: vllm/models/kimi_k3/nvidia/model.py [hunk 6]
```diff
@@ -775,6 +794,39 @@ def __init__(
...
             )
 
+    def merge_fused_front_weights(self) -> None:
+        down_proj = self.routed_expert_down_proj
+        if (
+            not self._fused_front_config_eligible
+            or self.shared_experts is None
+            or down_proj is None
+            or not isinstance(self.experts, LatentMoERunner)
+            or not self.experts._use_fused_path()
```
Comment (jeejeelee) [inline vllm/models/kimi_k3/nvidia/model.py:804]: I think we can merge the weights on different parallel config. is that right?

File: vllm/models/kimi_k3/nvidia/model.py [hunk 7]
```diff
@@ -775,6 +794,39 @@ def __init__(
...
+            or self.experts.do_naive_dispatch_combine
+            or self.experts.moe_config.pcp_size > 1
+            or self.experts.moe_config.hidden_dim != self.moe_hidden_size
+            or self.experts._quant_method.has_unpadded_output
+        ):
+            return
+        weights = [
+            self.shared_experts.gate_up_proj.weight,
+            self.gate.weight,
+            down_proj.weight,
```
Comment (jeejeelee) [inline vllm/models/kimi_k3/nvidia/model.py:814]: I think we should make sure all of them are unquantized weights

**Review by jeejeelee — CHANGES_REQUESTED** [review]
Thank you for this feature.  I think we should provide an  E2E accuracy and performance comparison here.

**Comment by yewentao256** [comment]
I don't think E2E show clear improvement, close the PR. Thanks for the review!
